### PR TITLE
Validate karaf features

### PR DESCRIPTION
**Pull request by gnodet**
Supercedes #43

### CODE REVIEW

**Comment by gnodet** [comment]
This PR fixes the features and provide build time validation of the features.
It ensures that the PR can actually be deployed by running the OSGi resolution at build time.

**Comment by bdemers** [comment]
Thanks!

**Comment by jbonofre** [comment]
R: @jbonofre

I'm resuming all work around Karaf 4 support and improvements on Shiro.

File: pom.xml [hunk 1]
```diff
@@ -76,8 +76,11 @@
         <!-- Don't change this version without also changing the shiro-aspect and shiro-features
              modules' OSGi metadata: -->
         <aspectj.version>1.8.9</aspectj.version>
+        <cas.client.core.version>3.2.2</cas.client.core.version>
         <commons.cli.version>1.3.1</commons.cli.version>
         <commons.codec.version>1.10</commons.codec.version>
+        <commons.configuration2.version>2.1</commons.configuration2.version>
```
Comment (steinarb) [inline pom.xml]: Version 2.2 of commons-configuration2 has the springbeans package imports as optional.

**Comment by fpapon** [comment]
@gnodet thanks for the PR!

I made a new PR based on this with an upgrade of Spring version.

https://github.com/apache/shiro/pull/88

I can close this one.

**Comment by steinarb** [comment]
Just FYI Re: shiro and karaf

It's not finished yet (as in deployed to maven central), but this is a minimal JDBC-based user system for karaf (currently using shiro 1.3.1): https://github.com/steinarb/authservice

Points of interest:

1. [uses code and OSGi service dependency injection to configure shiro](https://github.com/steinarb/authservice/blob/master/authservice.web.security/src/main/java/no/priv/bang/authservice/web/security/AuthserviceShiroFilter.java#L73) (shiro.ini can't be used, because the OSGi classloader is unable to find the shiro classes.  [The use of shiro.ini is limited to the [urls] section](https://github.com/steinarb/authservice/blob/master/authservice.web.security/src/main/resources/shiro.ini)). The injected OSGi services are realm and sessuib

2. [Uses liquibase to set up the schema and initial data](https://github.com/steinarb/authservice/blob/master/authservice.db.liquibase/src/main/resources/db-changelog/db-changelog-1.0.0.xml#L6), which means that it can easily be adapted to any JDBC database supported by liquibase (which is basically all of them), currently I'm using derby memory database for tests and PostgreSQL for production

3. Has a JDBC realm that supports base64 encoded salt (see [SHIRO-552](https://issues.apache.org/jira/browse/SHIRO-552)) (I'm working on a pull rquest for this, but it stranded on the OSGi issues that you are currently working)

4. has a very small GUI for logging in and letting users change their own passwords and email/real name (no actual frontend, just bootstrap-styled HTML delivered by jersey). There is also a jersey resource in this module that allows nginx to use shiro for authentication

5. has a react application for user management, the react application's jersey backend communicates with its business logic through [this OSGi service](https://github.com/steinarb/osgi-service/blob/master/osgiservice.users/src/main/java/no/priv/bang/osgiservice/users/UserManagementService.java#L33)

**Comment by fpapon** [comment]
@steinarb yes I know that the OSGi stuff on Shiro it's not finish yet ;)

We want to release the 1.4.1 soon and prepare the next 1.5.0 with a full OSGi integration.

All your points are very interesting and we can work together on it, I will start a thread on the mailing just after the release of the 1.4.1

**Pull request closed without merging**